### PR TITLE
SPIRVReader changes to handle opaque pointer transition

### PR DESCRIPTION
**Pull request by mariusz-sikora-at-amd**
This patch is introducing changes to functions like:
 - postProcessRowMajorMatrix
 - addLoadInstRecursively
 - addStoreInstRecursively
 - buildConstStoreRecursively
 - transValueWithOpcode<OpLoad>
 - transValueWithOpcode<OpStore>
 - transValueWithOpcode<OpAccessChain>
 - transValueWithOpcode<OpVariable>

Additionally one workaround was added to handle OpTypeForwardPointer
this workaround will be removed when transition is competed.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (kuhar) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:1733]: also here

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (kuhar) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:141]: Thanks for outlining this, I think it makes the callsites much cleaner.

nit: no need for an empty line before `return`

**Review by kuhar — COMMENTED** [review]
Just some cosmetic suggestions from me.

Have you checked what the impact on compilation times is? Historically, we found the translation code to have a few pathologically slow edge cases that only showed up on some large shaders.

**Comment by mariusz-sikora-at-amd** [comment]
> Just some cosmetic suggestions from me.
> 
> Have you checked what the impact on compilation times is? Historically, we found the translation code to have a few pathologically slow edge cases that only showed up on some large shaders.

Thanks for the comments. I will push new commit with changes.

I checked compilation time and the average time is  -0.1% (minus meaning faster).
The biggest slowdown is +0.91% (GeorgetownPipelineCs_0xB6464C22E9568A2A.pipe)
The biggest speedup is -1.15% (AshesPipelineVsFs_0x7B077AA1255F3655.pipe)

**Review by kuhar — DISMISSED** [review]
LGTM but please get a second approval before submitting

Also, remember to squash all commits before submitting. You can force-push to update the PR or use the Squash-and-merge button.

**Comment by mariusz-sikora-at-amd** [comment]
I squashed commits.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (kosarev) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: We don't seem to need vectors here at all? Can we just use initialiser list or `ArrayRef<>`?

Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: I'm running into some issues when using ArrayRef<> and "release" version of the code. It looks like "getBuilder()->getInt32" is not initializing object properly when invoked inside initialize_list.
Do you know about any limitations for ArrayRef<>

Comment (Flakebi) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: ArrayRef and Twine cannot be stored, so you cannot have a local variable of them (the backing memory immediately goes out of scope). It only works if you inline the list into the function call argument, like it was before.

Comment (kosarev) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: > It looks like "getBuilder()->getInt32" is not initializing object properly when invoked inside initialize_list.

That's odd. If I read the code right, all the elements point to values stored somewhere else, so it should be safe to pass an `ArrayRef<>` or initialiser list.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (piotrAMD) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:348]: s/complited/completed

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (piotrAMD) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:2884]: It would be more practical to use the same text in all the TODO's, so that it is easier to grep for all cases that need updating later (also something with "opaque" in the text would be helpful).

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (piotrAMD) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:2954]: Just something to keep in mind when you are working on other patches: this may need generalisation as I cannot convince myself that we only need to care about AccessChain as opposed to all instructions that return a pointer.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (piotrAMD) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:2951]: Seems unfortunate that we cannot unify transType and the new map, but I do get the argument about needing a dereferenced returned type in the new code.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: done

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: done

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: I returned to the initialization list for `CreateGEP` and added the same for `getIndexedType`. I also added comment because my intention was to have a single vector of indices and pass it to `CreateGEP` and `getIndexedType` which in my opinion is less error prone for later changes.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (tsymalla) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:2943]: nit: vectors

File: llpc/translator/lib/SPIRV/SPIRVReader.h
Comment (tsymalla) [inline llpc/translator/lib/SPIRV/SPIRVReader.h:295]: This is only adding in case the SPIRVValue * was not found inside the map, so I'd suggest to call this `tryAdAccessChainRetType` or something.

Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.h:295]: Good point. Will update

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:1376]: Yes.
I will update the comment. You are right `GEP` can be confusing.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (kosarev) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: Just in case: you can still do it your way by declaring `indices` an ordinary array, so `Value *indices[] = ...`, and passing it as an `ArrayRef<>`. Still no need for vectors.

**Comment by kuhar** [comment]
@mariusz-sikora-at-amd can you please close old conversations so that we know if there are outstanding issues or not? (Use the `Resolve conversation` button.)

**Comment by mariusz-sikora-at-amd** [comment]
Pushed one more update where I added support for OpSelect (if AccessChain is a condition of the OpSelect) and I moved all workarounds to a single function `getPointeeType`.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:716]: It meant to be a functional change for OpSelect when OpAccessChain is a input and it is pointing to a matrix. But it was based on wrong assumptions about variable pointers. Now we know that it is illegal to have a variable pointer as a object of OpTypeMatrix.
This change is no longer needed. I will remove it.
Thanks !

**Review by piotrAMD — DISMISSED** [review]

**Review by amdrexu — APPROVED** [review]

**Comment by mariusz-sikora-at-amd** [comment]
I rebased changes to trigger Jenkins.

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 757b0d8b945f674a5c50b09e731b8bd2a664bf8f
<details><summary>CTS tests (Failed: 0/177309)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35753/62859 (56.9%)</li>
<li>Failed: 0/62859 (0.0%)</li>
<li>Not Supported: 27106/62859 (43.1%)</li><li>Warnings: 0/62859 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 27845/51589 (54.0%)</li>
<li>Failed: 0/51589 (0.0%)</li>
<li>Not Supported: 23744/51589 (46.0%)</li><li>Warnings: 0/51589 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37998/62861 (60.4%)</li>
<li>Failed: 0/62861 (0.0%)</li>
<li>Not Supported: 24863/62861 (39.6%)</li><li>Warnings: 0/62861 (0.0%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Pull request merged**
